### ISSUE TITLE
Title: Issues with Evaluating Results using picai_eval in nnU-Net

### ISSUE THREAD

**Issue by hesam989** [DIAGNijmegen/picai_eval#15]
Hello,

I am currently working on a project using nnU-Net and have encountered an issue while trying to evaluate my results. I have successfully used nn-UNet for training and nnUNetv2_predict for predicting my test dataset. However, I am facing difficulties in properly evaluating the results, neither with Detection Maps nor with Softmax Volumes.

Here are the details of the issue:

Environment Details:

Path: e:\Hesam\envs\3dunetAghdam\lib\site-packages\sklearn\metrics\_ranking.py
Warning: UndefinedMetricWarning: No negative samples in y_true, false positive value should be meaningless
This warning suggests an issue with the evaluation metrics due to the absence of negative samples in y_true.
Evaluation Results:

For Detection Maps:
auroc: NaN
AP: 0.7681818181818182
num_cases: 43
num_lesions: 45
picai_eval_version: 1.4.x
lesion_results: Various values (e.g., 10021_1000021: [1, 1.0, 0.8476386663616025])
For Softmax Volumes:
auroc: NaN
AP: 0.8255997853760775
num_cases: 43
num_lesions: 45
picai_eval_version: 1.4.x
I am unsure why the auroc value is NaN and whether the warning about no negative samples is impacting the evaluation results. Any guidance or suggestions on how to resolve this issue would be greatly appreciated.

**Comment by joeranbosma** [comment]
Hi @hesam989,

It appears that the set of 43 cases you're evaluating on, are all positive cases (i.e., each case has at least one lesion in the ground truth label). Could that be the case? If so, the case-level diagnostic AUROC is NaN, because there are no negative samples.

Otherwise, the values appear as expected in `picai_eval`.

**Comment by mbu93** [comment]
I'd like to add to this question: I understood that case-level metrics require a set of lesion-free cases, obviously, as otherwise we can't determine TPs, FPs etc. For the lesion-level evaluation, however, including such lesion-free cases **significantly** reduces the AP achieved. I think the problem is that the following code:

```
evaluate(np.zeros((1, 2, 32, 32)), np.zeros((1, 2, 32, 32)))
```
results in an AP of 0. Isn't that counterintuitive? I'd expect a high metric (indeed, there IS no lesion). Or are we supposed to calculate this metric with cases that contain csPCa only? 
Furthermore: calculating the AP for a set of cases and then averaging the value, differs significantly from processing a batch as well.

**Comment by joeranbosma** [comment]
Hi @mbu93,

The average precision (AP) is a distribution metric, similar to the Area Under the Receiver Operating Characteristic Curve. As such, it is not well-defined for processing single cases. For more info on the AP, please check online resources (e.g., https://www.v7labs.com/blog/mean-average-precision).

In the provided example, the AP is 0 because the metric is not well defined, you need at least a single lesion. 

To calculate the correct AP, you should provide the **full set of cases, not single cases or batches**.

**Comment by mbu93** [comment]
Thanks for the swift response. That makes sense. Using a set of all-zero / all-one mock-predictions-label pairs, the metric returns 100%, as one would expect.

**Comment by joeranbosma** [comment]
Great! Does that completely solve your issue, or is there anything else?
